tools/slo-generator/slo_generator/exporters/bigquery.py

```diff
@@ -0,0 +1,259 @@
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+`bigquery.py`
+BigQuery exporter class.
+"""
+import json
+import logging
+import google.api_core
+from google.cloud import bigquery
+
+LOGGER = logging.getLogger(__name__)
+
+class BigqueryExporter(object):
+    """BigQuery exporter class."""
+
+    def __init__(self):
+        self.client = bigquery.Client(project="unset")
+
+    def export(self, data, **config):
+        """Export results to BigQuery.
+
+        Args:
+            data (dict): Data to export.
+                service_name (str): Service name.
+                feature_name (str): Feature name.
+                slo_name (str): SLO name.
+                timestamp_human (str): Timestamp in human-readable format.
+                measurement_window_seconds (int): Measurement window (in s).
+
+            config (dict): Exporter config.
+                project_id (str): BigQuery dataset project id.
+                dataset_id (str): BigQuery dataset id.
+                table_id (str): BigQuery table id.
+
+        Raises:
+            BigQueryError (object): BigQuery exception object.
+        """
+        project_id = config['project_id']
+        dataset_id = config['dataset_id']
+        table_id = config['table_id']
+        self.client.project = project_id
+        table_ref = self.client.dataset(dataset_id).table(table_id)
+        try:
+            table = self.client.get_table(table_ref)
+        except google.api_core.exceptions.NotFound:
+            table = self.create_table(
+                project_id,
+                dataset_id,
+                table_id,
+                schema=TABLE_SCHEMA)
+        row_ids = "%s%s%s%s%s" % (data["service_name"], data["feature_name"],
+                                  data["slo_name"], data["timestamp_human"],
+                                  data["window"])
+        results = self.client.insert_rows_json(
+            table,
+            json_rows=[data],
+            row_ids=[row_ids],
+            retry=google.api_core.retry.Retry(deadline=30)
+        )
+        if results != []:
+            raise BigQueryError(results)
+
+        return results
+
+    def create_table(self, project_id, dataset_id, table_id, schema=None):
+        """Creates a BigQuery table from a schema.
+
+        Args:
+            project_id (str): Project id.
+            dataset_id (str): Dataset id.
+            table_id (str): Table id to create.
+            schema (dict): BigQuery table schema in JSON format.
+        """
+        pyschema = []
+        if schema is not None:
+            schema = TABLE_SCHEMA
+        for row in schema:
```

Comment:
this assumes no nested schema which I think is fine in this case, would there be cases we'd need a nested schema?

Reply:
Not yet, but this will come as a subsequent PR (we need SLO metadata such as `owner`, `team`, etc...): what would differ in a nested schema here ?